interop-cats-effect/shared/src/main/scala/scalaz/zio/interop/bio/instances/ZioFiber2.scala

```diff
@@ -0,0 +1,40 @@
+package scalaz.zio
+package interop
+package bio
+package instances
+
+import cats.syntax.either._
+import cats.syntax.option._
+import scalaz.zio.Exit.{ Failure, Success }
+
+object ZioFiber2 {
+
+  @inline def fromFiber[E, A](fiber: Fiber[E, A]): Fiber2[IO, E, A] =
+    new Fiber2[IO, E, A] {
+
+      def await: IO[Nothing, Option[Either[E, A]]] =
+        fiber.await >>= fromExit
+
+      def cancel: IO[Nothing, Option[Either[E, A]]] =
+        fiber.interrupt >>= fromExit
+
+      def join: IO[E, A] =
+        fiber.join
+    }
+
+  private[this] def fromExit[E, A](exit: Exit[E, A]): IO[Nothing, Option[Either[E, A]]] =
+    exit match {
+      case Success(a) =>
+        IO.succeed(a.asRight.some)
+
+      case Failure(cause) =>
+        if (cause.failed)
+          IO.succeed(
+            cause.failures.headOption map { e: E =>
```

Comment:
This is clearly not ideal as we are losing errors here. The failure model of `Fiber2` doesn't look expressive enough to support the Zio `Fiber`'s. I was thinking to change the return type to something like
```scala
F[Nothing, Option[Either[FailedWith[E], A]]]
```
where `FailedWith` is on the line of
```scala
sealed trait FailedWith[+E] extends Product with Serializable

object FailedWith {

  final case class Single[E](failure: E)          extends FailedWith[E]
  final case class Multiple[E](failures: List[E]) extends FailedWith[E]
}
```
or even adding to `FailedWith` a
```scala
final case object Interrupted extends FailedWith[Nothing]
```
and get as a result of `await` and `cancel`
```scala
F[Nothing, Either[FailedWith[E], A]]
```
What do you think ?

Reply:
It's a big change. I can't see other effect types supporting it.